scRNA-seq-advanced/05-aucell.Rmd

```diff
@@ -0,0 +1,403 @@
+                                            assignCells = TRUE)
+```
+
+We're going to plot the distribution of AUC values with `ggplot2`, so we will want the AUC values in a longer format.
+
+```{r auc_plotting_df}
+auc_plotting_df <- auc_df |>
+  tidyr::pivot_longer(!barcodes,
+                      names_to = "gene_set",
+                      values_to = "auc") |>
+  dplyr::mutate(
+    # Create a new logical column called assigned
+    assigned = dplyr::case_when(
+      # For Zhang gene set rows, set to TRUE when the barcode is in the assignment list
+      gene_set == ewing_gene_set_names[1] & 
+        barcodes %in% auc_assignments[[ewing_gene_set_names[1]]]$assignment ~ TRUE,
+      # For Riggi gene set rows, set to TRUE when the barcode is in the 
+      # assignment list
+      gene_set == ewing_gene_set_names[2] & 
+        barcodes %in% auc_assignments[[ewing_gene_set_names[2]]]$assignment ~ TRUE,
+      # Otherwise, set to FALSE
+      .default = FALSE
+    )
+  )
+
+auc_plotting_df
+```
+
+To draw vertical lines representing the automatically chosen threshold, we can create a separate data frame.
+
+```{r auc_threshold_df}
+auc_threshold_df <- data.frame(
+  gene_set = ewing_gene_set_names,
+  # Grab thresholds associated with each gene set from assignements object
+  threshold = c(auc_assignments[[ewing_gene_set_names[1]]]$aucThr$selected, 
+                auc_assignments[[ewing_gene_set_names[2]]]$aucThr$selected)
+)
+
+auc_threshold_df
+```
+Now let's make a density plot, plotting the density of the assigned and unassigned cells separately and drawing a vertical line for the threshold.
+
+```{r auc_density_plot}
+auc_plotting_df |>
+  ggplot2::ggplot(
+    ggplot2::aes(
+      x = auc,  # AUC values
+      color = assigned,  # Group by assignment
+      fill = assigned,   # Group by assignment
+    )
+  ) +
+  ggplot2::geom_density(alpha = 0.2) +
+  # Draw a vertical dotted line showing the threshold for each gene set
+  ggplot2::geom_vline(data = auc_threshold_df,
+                      mapping = ggplot2::aes(xintercept = threshold),
+                      lty = 2) +
+  # Plot each gene set in its own facet
+  ggplot2::facet_grid(cols = ggplot2::vars(gene_set)) +
+  # Use a built-in theme
+  ggplot2::theme_bw()
+```
+
+For these particular gene sets, the AUC values appear to be bimodally distributed, and we can easily identify cells where the genes are highly expressed.
+
+### UMAPs
+
+#### `colData`
+
+We can also add the AUC values back into the SingleCellExperiment for convenience, e.g., for plotting.
+We'll add it to the existing `colData`.
+
+```{r coldata, live = TRUE}
+# Extract the existing colData, and left join it with the AUC values by the
+# barcodes
+coldata_df <- colData(sce) |>
+  as.data.frame() |>
+  dplyr::left_join(
+    auc_df |> 
+      # Use shorter names
+      dplyr::rename(zhang_auc = ewing_gene_set_names[1],
+                    riggi_auc = ewing_gene_set_names[2]),
+    by = "barcodes"
```

Comment:
If you rename from the beginning you could avoid the renaming here and in the other data frames you make.